evolvedbinary/prosemirror-lwdita
# Improve user UX regarding errors in the editor during file & GitHub operations

### Description

This PR will improve the UX by handling all errors that result from either erroneous parameters, uploading an invalid XDITA file, or receiving error responses from GitHub & server by redirecting to an error page.

Function `showErrorPage()`will take 3 arguments: `errorType`, `referer`, and `errorMessage`, which are going to be passed to the error page as parameters.

Currently I have implemented following errorTypes:
```
refererMissing
invalidParams
missingAuthentication
unknownError
fileNotFound
fileUploadError
```
 
The error page will display a set of markup and content depending on the received error-type, referer link, or error-message and dynamically handle all combinations of parameters for each error type. 

<img width="716" alt="Screenshot 2024-10-09 at 14 30 36" src="https://github.com/user-attachments/assets/c8b81c2e-40c1-42f3-8229-f870e05141fc">

<img width="551" alt="Screenshot 2024-10-09 at 16 52 12" src="https://github.com/user-attachments/assets/bd6890fd-cc4f-4fa0-a308-ab89ac5f2604">


### How to test

1. Manual test of the error page: Insert following URLs to test the displayed content. You can play around with the parameters, add or delete a value how you like:
   -  error type invalidParams: http://localhost:1234/error.html?error-type=invalidParams&referer=https://example.com
   -  error type refererMissing: http://localhost:1234/error.html?error-type=refererMissing
   -  error type missingAuthentication: http://localhost:1234/error.html?error-type=missingAuthentication
   -  error type missingAuthentication with referer: http://localhost:1234/error.html?error-type=missingAuthentication &referer=https://example.com
   - Missing authentication with error message: http://localhost:1234/error.html?error-type=missingAuthentication&referer=https://example.com&error-msg=a%20very%20bad%20error
   - unknown error: http://localhost:1234/error.html?error-type=unknownError&referer=https://example.com&error-msg=unknown%20error
   - File not found: http://localhost:1234/error.html?error-type=fileNotFound&referer= https://example.com&error-msg=Nofile

2. Test File Upload:

   - Go to the Petal start page and upload a file that contains an error, e.g. a missing close tag.
   - Check the redirect and the content of the page
   - Check the back-link to the start page

3. Test a round trip by publishing a file change, but provoke any errors at the point, where I have inserted the showErrorPage function in the Github plugin file to trigger the error page redirect.
